service/topology_coordinator.cc

```diff
@@ -1592,6 +1611,20 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
             co_return;
         }
 
+        // generate updates for pending resize finalizations
+        if (has_pending_resize_finalization && !utils::get_local_injector().enter("tablet_split_finalization_within_tablet_migration_postpone")) {
+            updates.clear();
+            guard = co_await generate_resize_finalization_updates(updates, std::move(guard));
+
+            // `generate_resize_finalization_updates` will generate updates iff there are no active transitions.
+            // so, if updates are not empty, we can continue or else, we need check for active transitions.
+            if (updates.empty() && has_tablet_transitions()) {
+                // transitions were started when `generate_resize_finalization_updates` attempted
```

Comment:
@tgrabiec @lkshminarayanan I don't follow this. for my education, how could we have transitions here, at this stage, if we handle it above in the has_transitions case (returning early before reaching here)? has_transitions should be the same as has_tablet_transitions() afaik.

Reply:
the migration tracker should be able to see all migrations emitted up to current point in time, right? it's only advancing migrations previously started. and afaik, balancer cannot run concurrently to migration tracker, so we don't need a read barrier to catch new migrations emitted. what am I missing?

Reply:
> @raphaelsc `generate_resize_finalization_updates()` executes a global metadata barrier and during that a concurrent transaction (like a `move_tablet()` API) can start new transitions. Previously that was not possible as we exclusively handled resize finalization within its own topology state but now since we are handling finalization inside `tablet_migration` state, we need to handle any new possible transitions after the global metadata barrier. I have also added an example that reproduces this case with a `move_tablet()` in parallel.

that's what I was missing. I didn't see your reply because my page was stale. I think we should add a comment to explain that. what you wrote above, with some rewording, can make a nice comment.

Reply:
Sure! I'll update the code. Thank you.